apps/designer/app/shared/router-utils/user-content-params.ts

```diff
@@ -0,0 +1,90 @@
+export type CanvasRouteMode = "edit" | "preview" | "published";
+const modes = ["edit", "preview", "published"] as CanvasRouteMode[];
+const getMode = (url: URL): CanvasRouteMode => {
+  const modeParam = url.searchParams.get("mode");
+  const mode =
+    modeParam === null ? "published" : modes.find((mode) => mode === modeParam);
+  if (mode === undefined) {
+    throw new Error(`Invalid mode "${modeParam}"`);
+  }
+  return mode;
+};
+
+// A subtype of Request. To make testing easier.
+type MinimalRequest = {
+  url: string;
+  headers: { get: (name: string) => string | null };
+};
+
+const getRequestHost = (request: MinimalRequest): string =>
+  request.headers.get("x-forwarded-host") || request.headers.get("host") || "";
+
+export const getUserContentBaseUrl = (request: MinimalRequest): string => {
+  const { USER_CONTENT_BASE_URL } = process.env;
+  if (USER_CONTENT_BASE_URL !== undefined && USER_CONTENT_BASE_URL !== "") {
+    return USER_CONTENT_BASE_URL;
+  }
+
+  // Local development special case
+  const host = getRequestHost(request);
+  if (
+    process.env.NODE_ENV === "development" &&
+    /^(.*\.)?localhost(:\d+)?$/.test(host)
```

Comment:
why not new URL(host).hostname?

Reply:
I don't understand. This code checks that `host` looks like `"localhost"` or `"foo.localhost"` or `"foo.localhost:3000"`.

Reply:
oh I see, I just wish we would avoid regexes if we can use some well-tested parser, at least add the examples please to make it clear what its for